src/connector/src/sink/file_sink/opendal_sink.rs

```diff
@@ -195,61 +213,156 @@ pub struct OpenDalSinkWriter {
-                    let _ = w.close().await?;
+                    w.close().await?;
                 }
             };
         }
+        Ok(false)
+    }
 
-        Ok(())
+    /// Attempts to complete the file write using the writer's internal batching strategy.
+    ///
+    /// If the write is completed, returns the last `chunk_id` that was written;
+    /// otherwise, returns None.
+    ///
+    /// For the file sink, currently, the sink decoupling feature is not enabled.
+    /// When a checkpoint arrives, the force commit is performed to write the data to the file.
+    /// In the future if flush and checkpoint is decoupled, we should enable sink decouple accordingly.
+    pub async fn try_finish(&mut self) -> Result<Option<usize>> {
+        match self.try_finish_write_via_rollover_interval().await? {
+            true => Ok(self.written_chunk_id),
+            false => Ok(None),
+        }
+    }
+
+    /// This asynchronous function checks if the duration since the writer was
+    /// created exceeds the configured rollover interval (`rollover_seconds`).
+    /// If the duration condition is met and a sink writer is available, it
+    /// closes the current writer (e.g., a Parquet file writer) if any bytes
+    /// have been written. The function logs the completion message and
+    /// increments the writer index for subsequent writes.
+    ///
+    ///
+    /// Returns:
+    /// - `Ok(true)` if the write was finalized successfully due to the
+    ///   rollover interval being exceeded.
+    /// - `Ok(false)` if the rollover interval has not been met, indicating that
+    ///   no action was taken.
+    async fn try_finish_write_via_rollover_interval(&mut self) -> Result<bool> {
+        if self.duration_seconds_since_writer_created()
+            >= self.batching_strategy.clone().rollover_seconds
+            && self.commit().await?
+        {
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    /// This asynchronous function checks if the number of rows accumulated
+    /// (`current_bached_row_num`) meets or exceeds the maximum row count
+    /// defined in the `batching_strategy`. If the threshold is met, it
+    /// closes the current writer (e.g., a Parquet file writer), logs the
+    /// completion message, and resets the row count. It also increments
+    /// the writer index for subsequent writes.
+    ///
+    /// Returns:
+    /// - `Ok(true)` if the write was finalized successfully.
+    /// - `Ok(false)` if the threshold has not been met, indicating that
+    ///   no action was taken.
+    async fn try_finish_write_via_batched_rows(&mut self) -> Result<bool> {
+        if self.current_bached_row_num >= self.batching_strategy.clone().max_row_count
+            && self.commit().await?
+        {
+            self.current_bached_row_num = 0;
+            return Ok(true);
+        }
+
+        Ok(false)
+    }
+
+    pub async fn should_finish(&mut self) -> Result<Option<usize>> {
```

Comment:
We have both `commit` and `finish` representing the same thing in this file. Let's unify the terminology to use only `finish` or only `commit` in the naming instead.

Reply:
The two interfaces are a bit different, through they both do the close writer logic, `commit` return bool to represent whether writer is successfully closed, while `should_finish` return `chunk_id`, which is used when truncating log store.
I'd like to change `commit` from public to private as it's used in `try_finish_xxx` and `should_finish`, and leave the `finish` interfaces public to `BatchingLogSinker`. Is the acceptable for you?

Reply:
Let me give more details based on the [comment ](https://github.com/risingwavelabs/risingwave/pull/18472#discussion_r1793084847) here to organize the interfaces better (I will use the commit to replace finish to simplify the example):
```rust
impl OpenDalSinkWriter {
    // Public
    // Put data into the writer
    pub async fn write_batch(...) -> Result<()> {
       ...
       writer.write(...)
       // Update current_bached_row_num
       ...
    }
    
    // Public
    // Trigger a commit (i.e. close writher)
    pub async fn commit(...) ->Result<ChunkId>. {
        if writer.bytes_written() > 0 {
            writer.close().await?;
            ...
        }
    }
   
   // Public
   // Try commit if the batching condition is met
   pub fn try_commit(...) -> Result<Option<ChunkId>> {
        if can_commit(...) {
            self.commit(...)
        }
        ...
    }
    
    // Private
    // Methods to whether batching is finished
    fn can_commit(...) -> bool {
        self.exceed_batch_row_count() || self.exceed_rollover_interval()
    }

    // Private.
   // We can also put the logics inside can_commit without defining these methods
   fn exceed_batch_row_count(...) -> bool {}
   fn exceed_rollover_interval(...) -> bool {}

}
```

In the above example, write batch, commit, and batching strategy are completely independently (`try_commit` still couples commit and batching strategy, which is fine because it is just a util method. The caller can check `can_commit` and call `commit` directly without using this method technically). And here is a cleaner example on how batch log sinker will use the public methods:
```rust
// When scale happens
writer.commit(...)

// On receiving a chunk
writer.write_batch(...)
writer.try_commit(...)

// On receiving a barrier
writer.try_commit(...)
```

Note that in the future for exactly once support, you can still use the combination of `writer_batch`, `commit`, `try_commit` to fullfill the needs without introducing more public methods.

Reply:
Fully understand.
Previously, I always couple `write_batch` and `finish` in one interface `write_batch_and_try_finish()` because the `chunk_id` of the write can only be obtained when writing a chunk, and we should return chunk_id for truncating when closing writer. But in this way, `chunk_id` cannot be updated when receiving barrier and then committing.

Therefore, it is changed to store two chunk_id:  `finish_commit_chunk_id` and `written_chunk_id` in the writer, and then
 
- update `written_chunk_id` when `writing_batch`,
- update `finish_commit_chunk_id`=`written_chunk_id` when commit succeeds.

Thus commit and write_batch can be decoupled, batching and write_batch is also decoupled.